pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/IngestionDelayTracker.java

```diff
@@ -0,0 +1,371 @@
+ *   timeoutInactivePartitions()    stopTrackingPartitionIngestionDelay(partitionId)
+ *    _________|__________                \
+ *   | TimerTrackingTask |          (CONSUMING -> DROPPED state change)
+ *   |___________________|
+ *
+ */
+
+public class IngestionDelayTracker {
+
+  // Sleep interval for timer thread that triggers read of ideal state
+  private static final int TIMER_THREAD_TICK_INTERVAL_MS = 300000; // 5 minutes +/- precision in timeouts
+  // Once a partition is marked for verification, we wait 10 minutes to pull its ideal state.
+  private static final int PARTITION_TIMEOUT_MS = 600000;          // 10 minutes timeouts
+  // Delay Timer thread for this amount of time after starting timer
+  private static final int INITIAL_TIMER_THREAD_DELAY_MS = 100;
+
+  /*
+   * Class to keep an ingestion delay measure and the time when the sample was taken (i.e. sample time)
+   * We will use the sample time to increase ingestion delay when a partition stops consuming: the time
+   * difference between the sample time and current time will be added to the metric when read.
+   */
+  static private class DelayMeasure {
+    public DelayMeasure(long t, long d) {
+      _delayMilliseconds = d;
+      _sampleTime = t;
+    }
+    public final long _delayMilliseconds;
+    public final long _sampleTime;
+  }
+
+  // HashMap used to store delay measures for all partitions active for the current table.
+  private final ConcurrentHashMap<Integer, DelayMeasure> _partitionToDelaySampleMap = new ConcurrentHashMap<>();
+  // We mark partitions that go from CONSUMING to ONLINE in _partitionsMarkedForVerification: if they do not
+  // go back to CONSUMING in some period of time, we confirm whether they are still hosted in this server by reading
+  // ideal state. This is done with the goal of minimizing reading ideal state for efficiency reasons.
+  private final ConcurrentHashMap<Integer, Long> _partitionsMarkedForVerification = new ConcurrentHashMap<>();
+
+  final int _timerThreadTickIntervalMs;
+  // Timer task to check partitions that are inactive against ideal state.
+  private final Timer _timer;
+
+  private final ServerMetrics _serverMetrics;
+  private final String _tableNameWithType;
+  private final String _metricName;
+
+  private final boolean _enablePerPartitionMetric;
+  private final boolean _enableAggregateMetric;
+  private final Logger _logger;
+
+  private final RealtimeTableDataManager _realTimeTableDataManager;
+  private Clock _clock;
+
+  /*
+   * Returns the maximum ingestion delay amongs all partitions we are tracking.
+   */
+  private DelayMeasure getMaximumDelay() {
+    DelayMeasure newMax = null;
+    for (int partitionGroupId : _partitionToDelaySampleMap.keySet()) {
+      DelayMeasure currentMeasure = _partitionToDelaySampleMap.get(partitionGroupId);
+      if ((newMax == null)
+          ||
+          (currentMeasure != null) && (currentMeasure._delayMilliseconds > newMax._delayMilliseconds)) {
```

Comment:
I thought there was a comment to conform to `Ms` everywhere instead of "millis" or "millseconds", etc.?

Reply:
Thanks for pointing this out, there is no clear convention through the full code but it is good to keep consistency through one file at least, addressed in commit ecfcf1b1cdf81f4fe3547e1eec4ac23a1df72c5c